pkg/cache/sql/store/store.go

```diff
@@ -0,0 +1,370 @@
+)
+
+// Store is a SQLite-backed cache.Store
+type Store struct {
+	c       cache.Indexer
+	name    string
+	encrypt bool
+	typ     reflect.Type
+	keyFunc cache.KeyFunc
+	dbLock  *sync.RWMutex
+
+	DBClient
+
+	shouldEncrypt bool
+
+	upsertStmt   *sql.Stmt
+	deleteStmt   *sql.Stmt
+	getStmt      *sql.Stmt
+	listStmt     *sql.Stmt
+	listKeysStmt *sql.Stmt
+
+	afterUpsert []func(key string, obj any, tx db.TXClient) error
+	afterDelete []func(key string, tx db.TXClient) error
+}
+
+// Test that Store implements cache.Indexer
+var _ cache.Store = (*Store)(nil)
+
+type DBClient interface {
+	Begin() (db.TXClient, error)
+	Prepare(stmt string) *sql.Stmt
+	QueryForRows(ctx context.Context, stmt transaction.Stmt, params ...any) (*sql.Rows, error)
+	ReadObjects(rows db.Rows, typ reflect.Type, shouldDecrypt bool) ([]any, error)
+	ReadStrings(rows db.Rows) ([]string, error)
+	Upsert(tx db.TXClient, stmt *sql.Stmt, key string, obj any, shouldEncrypt bool) error
+	CloseStmt(closable db.Closable) error
+	// TxModeClient() DBClient
+}
+
+var backoffRetry = wait.Backoff{Duration: 50 * time.Millisecond, Factor: 2, Steps: 10}
+
+// NewStore creates a SQLite-backed cache.Store for objects of the given example type
+func NewStore(example any, keyFunc cache.KeyFunc, c DBClient, shouldEncrypt bool, name string) (*Store, error) {
+	s := &Store{
+		name:          sanitize(name),
+		typ:           reflect.TypeOf(example),
+		DBClient:      c,
+		keyFunc:       keyFunc,
+		shouldEncrypt: shouldEncrypt,
+		afterUpsert:   []func(key string, obj any, tx db.TXClient) error{},
+		afterDelete:   []func(key string, tx db.TXClient) error{},
+	}
+
+	// once multiple informerfactories are needed, this can accept the case where table already exists error is received
+	txC, err := s.Begin()
+	if err != nil {
+		return nil, err
+	}
+	err = txC.Exec(fmt.Sprintf(`CREATE TABLE "%s" (
+		key VARCHAR UNIQUE NOT NULL PRIMARY KEY,
+		object BLOB, objectnonce VARCHAR, dek VARCHAR, deknonce VARCHAR
+	)`, s.name))
+	if err != nil {
+		return nil, err
+	}
+
+	err = txC.Commit()
+	if err != nil {
+		return nil, err
+	}
+	s.upsertStmt = s.Prepare(fmt.Sprintf(upsertStmtFmt, s.name)) // ON CONFLICT DO UPDATE SET object = excluded.object`, s.name))
+	s.deleteStmt = s.Prepare(fmt.Sprintf(deleteStmtFmt, s.name))
+	s.getStmt = s.Prepare(fmt.Sprintf(getStmtFmt, s.name))
+	s.listStmt = s.Prepare(fmt.Sprintf(listStmtFmt, s.name))
+	s.listKeysStmt = s.Prepare(fmt.Sprintf(listKeysStmtFmt, s.name))
+
+	return s, nil
+}
+
+/* Core methods */
+// TODO: this is where I left off, need to figure out how lock will be handled
+// upsert saves an obj with its key, or updates key with obj if it exists in this Store
+func (s *Store) upsert(key string, obj any) error {
+	tx, err := s.Begin()
+	if err != nil {
+		return err
+	}
+
+	err = s.Upsert(tx, s.upsertStmt, key, obj, s.shouldEncrypt)
+	if err != nil {
+		return err
+	}
+
+	err = s.runAfterUpsert(key, obj, tx)
+	if err != nil {
+		return err
+	}
```

Comment:
I agree, this needs refactoring.

Either the registered function has the explicit duty of rolling back the transaction, or this code should do that.

Currently it's mixed: if the called code does (eg. by using the provided tx.StmtExec func) then it will perform the rollback, if failing for other reasons it won't.

-> [epic](https://github.com/rancher/rancher/issues/45635)